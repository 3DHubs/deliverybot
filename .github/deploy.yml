production:
  auto_deploy_on: refs/heads/master
  environment: production
  production_environment: true
  url: 'https://deliverybot-lhanl2ut4q-uc.a.run.app/'

  exec:
    image: cloud-run
    params:
      service: deliverybot
      platform: managed
      region: us-central1
      image:
        repository: gcr.io/coldog-181700/deliverybot
        tag: '{{short_sha}}'


  required_contexts:
  - "Build (390495068715)"

review:
  environment: pr{{ pr }}
  transient_environment: true
  production_environment: false

  exec:
    image: cloud-run
    params:
      service: deliverybot-pr{{pr}}
      platform: managed
      region: us-central1
      image:
        repository: gcr.io/coldog-181700/deliverybot
        tag: '{{short_sha}}'

  required_contexts:
  - "Build (390495068715)"
