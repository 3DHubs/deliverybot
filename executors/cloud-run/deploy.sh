#!/bin/sh
set -e

curl --silent --show-error --fail ${SECRETS_FILE} | jq -re '.GCLOUD_SERVICE_KEY' > service-key.json
gcloud auth activate-service-account --key-file=./service-key.json
echo $PARAMS > params.json

project=$(jq -r '.project_id' < service-key.json)
service="$(jq -r '.service' < params.json)"
platform="$(jq -r '.platform' < params.json)"
region="$(jq -r '.region' < params.json)"
image_repo="$(jq -r '.image.repository' < params.json)"
image_tag="$(jq -r '.image.tag' < params.json)"

gcloud beta run deploy "$service" \
  --project "$project" \
  --allow-unauthenticated \
  --platform "${platform:-managed}" \
  --region "${region:-us-central1}" \
  --image "${image_repo}:${image_tag}"
