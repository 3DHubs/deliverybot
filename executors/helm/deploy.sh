#!/bin/bash
set -e
curl --silent --show-error --fail ${SECRETS_FILE} | jq -re '.KUBECONFIG' > ./kubeconfig
export KUBECONFIG="$PWD/kubeconfig"

echo $PARAMS > params.json

namespace="$(jq '.namespace' < params.json)"
release="$(jq '.release' < params.json)"
chart="$(jq '.chart' < params.json)"

if [ "$namespace" = "" ]; then
  echo "'namespace' must be set in exec.params"
  exit 1
fi
if [ "$release" = "" ]; then
  echo "'release' must be set in exec.params"
  exit 1
fi
if [ "$chart" = "" ]; then
  chart="/charts/default"
fi

helm init --client-only
helm upgrade "$release" "$chart"  \
  --install --wait \
  --namespace "$namespace" \
  --values ./params.json
