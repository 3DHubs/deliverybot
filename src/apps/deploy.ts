import { AuthedRequest, authenticate, verifyRepo } from "./auth";
import { Response, Application } from "express";
import { request as ghRequest } from "@octokit/request";
import { deployCommit, config, Targets } from "./commands";
import * as pkg from "../package";
import { writeFileSync } from "fs";

export async function create(req: AuthedRequest, res: Response) {
  const { owner, repo, target, sha } = req.params;
  try {
    const deployed = await deployCommit(req.user!.github, (req as any).log, {
      owner,
      repo,
      target,
      sha,
      // TODO: Change to the branch here as well so that we can query by branch
      // later on.
      ref: sha
    });
    res.json({ message: "ok", sha, ...deployed.data });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
}

async function tryConfig(req: AuthedRequest) {
  const { owner, repo, branch } = req.params;
  try {
    return await config(req.user!.github, {
      owner,
      repo,
      ref: branch ? `refs/heads/${branch}` : `refs/heads/master`
    });
  } catch (err) {
    return null;
  }
}

export async function get(req: AuthedRequest, res: Response) {
  const { owner, repo, target, branch } = req.params;
  const conf = await tryConfig(req);
  const targets = Targets(target, conf);

  const { branches, commits } = await commitQuery(
    req.user!.token,
    owner,
    repo,
    branch || "master"
  );
  const params = {
    target,
    targets,
    owner,
    repo,
    pkg,
    branch,
    branches,
    commits
  };
  res.render("deploy", params);
}

export async function redirect(req: AuthedRequest, res: Response) {
  const { owner, repo } = req.params;
  const conf = await tryConfig(req);
  const target = Object.keys(conf || {})[0] || "none";
  const branch = "master";
  res.redirect(`/deploy/${target}/${owner}/${repo}/${branch}`);
}

export function deploy(app: Application) {
  app.get("/:owner/:repo", authenticate, verifyRepo, redirect);
  app.get("/deploy/:owner/:repo", authenticate, verifyRepo, redirect);
  app.get(
    "/deploy/:target/:owner/:repo/:branch",
    authenticate,
    verifyRepo,
    get
  );
  app.post(
    "/deploy/:target/:owner/:repo/:sha",
    authenticate,
    verifyRepo,
    create
  );
}

const query = `
query commits(
  $owner: String!, $repo: String!, $branch: String!
) {
  repository(owner:$owner,name:$repo) {
    refs(refPrefix:"refs/heads/",first:50) { nodes { name } }
    ref(qualifiedName:$branch) {
      target {
        ... on Commit {
          history(first: 10) {
            edges {
              node {
                messageHeadline oid message
                author { user { login } }
                status { contexts { state } }
                checkSuites(last: 50) { nodes { conclusion status } }
                deployments(last: 50) {
                  nodes {
                    id environment
                    creator { login }
                    latestStatus { logUrl state }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}`;

async function commitQuery(
  token: string,
  owner: string,
  repo: string,
  branch: string
) {
  const resp = await ghRequest({
    data: { query, variables: { owner, repo, branch } },
    method: "POST",
    url: "/graphql",
    headers: {
      authorization: `token ${token}`,
      accept: "application/vnd.github.antiope-preview+json"
    }
  });
  if (resp.data.errors && resp.data.errors.length > 0) {
    throw new Error(
      `Graphql request failure: ${JSON.stringify(resp.data.errors)}`
    );
  }
  writeFileSync('query.json', JSON.stringify(resp.data));
  return View(owner, repo, branch, resp.data.data);
}

export function View(owner: string, repo: string, branch: string, data: any) {
  const branches = Branches(branch, data.repository.refs.nodes);
  const commits = data.repository.ref.target.history.edges.map((edge: any) =>
    Commit(owner, repo, edge)
  );
  return { commits, branches };
}

const failing = ["FAILURE", "FAILING", "ERROR", "CANCELLED", "TIMED_OUT"];
const pending = ["PENDING", "ACTION_REQUIRED", "QUEUED", "IN_PROGRESS"];
const success = ["SUCCESS"];

function AggregateStatus(statuses: string[]) {
  if (statuses.find(s => failing.includes(s.toUpperCase()))) {
    return "FAILURE";
  }
  if (statuses.find(s => pending.includes(s.toUpperCase()))) {
    return "PENDING";
  }
  if (statuses.find(s => success.includes(s.toUpperCase()))) {
    return "SUCCESS";
  }
  return "WAITING";
}

function Check(node: any) {
  const statuses: string[] = ((node.status && node.status.contexts) || [])
    .map((ctx: any) => ctx.state)
    .concat(
      node.checkSuites.nodes.map((cs: any) => cs.conclusion || cs.status)
    );
  return { status: Status(AggregateStatus(statuses)) };
}

function Status(status?: string) {
  const s = [(status || "WAITING").toUpperCase()];
  switch (AggregateStatus(s)) {
    case "FAILURE":
      return { failure: true, color: "red" };
    case "PENDING":
      return { pending: true, color: "yellow" };
    case "SUCCESS":
      return { success: true, color: "green" };
    default:
      return { waiting: true, color: "gray" };
  }
}

function Deployment(node: any) {
  const statuses = (node.deployments || []).nodes.map(
    (deploy: any) =>
      (deploy.latestStatus && deploy.latestStatus.state) || "WAITING"
  );
  return { status: Status(AggregateStatus(statuses)) };
}

function Undeployed(node: any) {
  const deployment = Deployment(node)
  const check = Check(node)
  if (check.status.success && deployment.status.waiting) {
    return true
  }
  return false
}

function Deployments(node: any) {
  return (node.deployments || []).nodes.map((deploy: any) => ({
    status: Status(deploy.latestStatus && deploy.latestStatus.state),
    environment: deploy.environment,
    creator: deploy.creator.login,
    url: deploy.latestStatus && deploy.latestStatus.logUrl
  }));
}

function Commit(owner: string, repo: string, edge: any) {
  return {
    owner,
    repo,
    message: edge.node.messageHeadline,
    oid: edge.node.oid.substr(0, 7),
    author: edge.node.author.user.login,
    undeployed: Undeployed(edge.node),
    deployment: Deployment(edge.node),
    deployments: Deployments(edge.node),
    check: Check(edge.node)
  };
}

function Branches(
  active: string,
  branches: any[]
): Array<{ name: string; active: boolean }> {
  return branches.map(n => ({
    name: n.name,
    active: active === n.name
  }));
}

function Targets(target: string, conf: Targets | null) {
  return (
    conf &&
    Object.keys(conf).map((name: string) => ({
      name,
      active: name === target
    }))
  );
}
