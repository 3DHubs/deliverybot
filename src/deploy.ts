export { config, deployCommit as deploy } from "./app";
