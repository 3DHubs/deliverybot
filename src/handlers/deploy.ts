import { Application, Context } from "probot";
import { DeployState, run, Action, Deploy } from "../execute";
import { WebhookPayloadDeploymentDeployment } from "@octokit/webhooks";

const preview = "application/vnd.github.ant-man-preview+json";

function withPreview<T>(arg: T): T {
  (arg as any).headers = { accept: preview };
  return arg as T;
}

function getDeploy(
  action: Action,
  token: string,
  repo: {
    repoId: string;
    owner: string;
    repo: string;
  },
  deployment: WebhookPayloadDeploymentDeployment
): Deploy {
  const payload = deployment.payload as any;
  return {
    ...repo,
    action,
    token,
    deploymentId: `${deployment.id}`,
    environment: deployment.environment,
    sha: deployment.sha,
    url: payload.url,
    exec: payload.exec
  };
}

class Handler {
  app: Application;
  context: Context;
  repo: { owner: string; repo: string; repoId: string };

  constructor(app: Application, context: Context) {
    this.app = app;
    this.context = context;
    this.repo = {
      repoId: `${this.context.payload.repository.id}`,
      owner: this.context.payload.repository.owner.login,
      repo: this.context.payload.repository.name
    };
  }

  async deploy(
    action: Action,
    deployment: WebhookPayloadDeploymentDeployment
  ): Promise<DeployState | null> {
    if (!(deployment.payload as any).exec) {
      this.context.log.info({ deployment }, "Deploy: not deploying, no exec");
      return null;
    }
    const token = await this.app.app.getInstallationAccessToken({
      installationId: this.context.payload.installation.id
    });
    this.context.log.info({ deployment }, "Deploy: executing run");
    return run(getDeploy(action, token, this.repo, deployment));
  }

  async setStatus(
    state: "error" | "failure" | "pending" | "success",
    url?: string,
    description?: string
  ) {
    this.context.log.info({ state, url, description }, "Deploy: set status");
    await this.context.github.repos.createDeploymentStatus(
      withPreview({
        ...this.repo,
        deployment_id: this.context.payload.deployment.id,
        target_url: url,
        state,
        description
      })
    );
    await this.context.github.repos.createStatus({
      ...this.repo,
      sha: this.context.payload.deployment.sha,
      description,
      state,
      log_url: url,
      context: `deploy/${this.context.payload.deployment.environment}`
    } as any);
  }

  async handleDeploy() {
    try {
      const deployment = this.context.payload.deployment;
      this.context.log.info({ deployment }, "Deploy: handling");
      const result = await this.deploy(Action.Deploy, deployment);
      if (!result) {
        return;
      }
      this.context.log.info({ result }, "Deploy: started");
      await this.setStatus(result.state, result.logs, result.description);
    } catch (err) {
      this.context.log.error({ error: err }, "Deploy: error");
      await this.setStatus("failure");
    }
  }

  async handlePRClose(): Promise<void> {
    const ref = this.context.payload.pull_request.head.ref;
    const deployments = await this.context.github.repos.listDeployments(
      withPreview({ ...this.repo, ref })
    );
    for (const deployment of deployments.data) {
      // Only terminate transient environments.
      if (!deployment.transient_environment) {
        continue;
      }
      if (!deployment.payload || !(deployment.payload as any).exec) {
        continue;
      }
      this.context.log.info({ deployment }, "Deploy: removing");
      // Store all transient environments in a set.
      await this.context.github.repos.createDeploymentStatus(
        withPreview({
          ...this.repo,
          deployment_id: deployment.id,
          state: "inactive"
        })
      );
      // Delete the deployment.
      try {
        await this.deploy(Action.Remove, deployment as any);
      } catch (error) {
        this.context.log.error(
          { error: error.message },
          "Deploy: removal failed"
        );
      }
    }
  }
}

export function deploy(app: Application) {
  app.on("deployment", async context => {
    const handler = new Handler(app, context);
    await handler.handleDeploy();
  });
  app.on("pull_request.closed", async context => {
    const handler = new Handler(app, context);
    await handler.handlePRClose();
  });
}
