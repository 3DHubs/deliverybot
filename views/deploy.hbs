{{> repo_header}}

<nav class="UnderlineNav">
  <div class="content">
    <div class="UnderlineNav-body">
      <a href="/deploy/{{owner}}/{{repo}}" role="tab" title="Deploy" class="UnderlineNav-item selected">Deploy</a>
    </div>
  </div>
</nav>
</header>

<div class="bg-white fill">
  <div class="content">
    <div id="error-container"></div>
    <details class="details-reset details-overlay pt-3" style="display: inline-block;">
      <summary class="btn btn-sm" type="button" aria-haspopup="true" aria-expanded="true">
        Target: <b>{{target}}</b>
      </summary>
      <div class="SelectMenu">
        <div class="SelectMenu-modal">
          <menu class="SelectMenu-list">
            {{#each targets}}
            <a href="/deploy/{{name}}/{{../owner}}/{{../repo}}/{{../branch}}" class="SelectMenu-item" role="menuitem">
              {{name}}
            </a>
            {{/each}}
          </menu>
        </div>
      </div>
    </details>

    <details class="details-reset details-overlay pt-3" style="display: inline-block;">
      <summary class="btn btn-sm" type="button" aria-haspopup="true" aria-expanded="true">
        Branch: <b>{{branch}}</b>
      </summary>
      <div class="SelectMenu">
        <div class="SelectMenu-modal">
          <menu class="SelectMenu-list">
            {{#each branches}}
            <a href="/deploy/{{../target}}/{{../owner}}/{{../repo}}/{{name}}" class="SelectMenu-item" role="menuitem">
              {{name}}
            </a>
            {{/each}}
          </menu>
        </div>
      </div>
    </details>

    <div class="Box mt-3" id="commits">
      {{#each commits}}
      <div class="Box-row d-flex flex-items-center Box-row--hover-blue">
        <div class="flex-auto">
          <a class="text-gray-dark" target="_blank" href="https://github.com/{{owner}}/{{repo}}/commit/{{oid}}">
            {{message}}
          </a>
          <div class="text-gray-light">
            <strong><code>{{oid}}</code></strong> - <a href="https://github.com/{{author}}">@{{author}}</a>
          </div>
        </div>

        <div class="position-relative">
          <div id="commit-{{oid}}" class="Popover position-absolute d-none" style="right: 100%; bottom: -50%;">
            <div class="Popover-message Popover-message--large Popover-message--right-bottom mr-2 Box box-shadow-large Box--condensed">
              <div class="Box-header bg-white">
                <h5 class="text-{{check.status.color}} mb-1 mt-2">
                  {{#if check.status.success}}All checks have succeeded{{/if}}
                  {{#if check.status.pending}}Waiting for an in-progress check{{/if}}
                  {{#if check.status.failure}}Checks have failed{{/if}}
                  {{#if check.status.waiting}}Waiting for a check to be reported{{/if}}
                </h5>
                {{#if check.status.success}}
                  <p class="text-gray-light">Commit is ready for deployment</p>
                {{/if}}
                <div class="mt-4 mb-2">
                  <a target="_blank" href="https://github.com/{{owner}}/{{repo}}/commits/{{oid}}">
                    View commit status checks on GitHub
                  </a>

                </div>
             </div>
            </div>
          </div>
          <div style="margin-right: 16px;">
            <a href="#commit-{{oid}}" class="text-{{check.status.color}}" data-toggle="commit-{{oid}}">
              Commit
              {{#if check.status.success}}{{> icon_check}}{{/if}}
              {{#if check.status.pending}}{{> icon_dot}}{{/if}}
              {{#if check.status.failure}}{{> icon_x}}{{/if}}
              {{#if check.status.waiting}}{{> icon_clock}}{{/if}}
            </a>
          </div>
        </div>

        <div class="position-relative">
          <div id="deploy-{{oid}}" class="Popover position-absolute d-none" style="right: 100%; bottom: -50%;">
            <div class="Popover-message Popover-message--large Popover-message--right-bottom mr-2 Box box-shadow-large Box--condensed">
              <div class="Box-header bg-white">
                <h5 class="mb-2 mt-2 text-{{deployment.status.color}}">
                  {{#if deployment.status.success}}All deployments have succeeded{{/if}}
                  {{#if deployment.status.pending}}Waiting for an in-progress deployment{{/if}}
                  {{#if deployment.status.failure}}Deployments have failed{{/if}}
                  {{#if deployment.status.waiting}}Waiting for a deployment to be reported{{/if}}
                </h5>
              </div>
              {{#each deployments}}
              <div class="Box-row">
                <span class="text-{{status.color}} mr-2">
                  {{#if status.success}}{{> icon_check}}{{/if}}
                  {{#if status.pending}}{{> icon_dot}}{{/if}}
                  {{#if status.failure}}{{> icon_x}}{{/if}}
                  {{#if status.waiting}}{{> icon_clock}}{{/if}}
                </span>
                <strong>{{environment}}</strong>
                by {{creator}}
                {{#if url}}
                <a href="{{url}}" class="float-right">Details</a>
                {{/if}}
              </div>
              {{/each}}
            </div>
          </div>
          <div style="margin-right: 16px;">
            <a href="#deploy-{{oid}}" class="text-{{deployment.status.color}}" data-toggle="deploy-{{oid}}">
              Deploy
              {{#if deployment.status.success}}{{> icon_check}}{{/if}}
              {{#if deployment.status.pending}}{{> icon_dot}}{{/if}}
              {{#if deployment.status.failure}}{{> icon_x}}{{/if}}
              {{#if deployment.status.waiting}}{{> icon_clock}}{{/if}}
            </a>
          </div>
        </div>


        {{#if ../target}}
        {{#if undeployed}}
        <button type="button" data-oid="{{oid}}" aria-label="This commit is ready for deployment to {{../target}}."
          class="deploy btn btn-primary tooltipped tooltipped-n"
          name="button">
          {{> icon_play}}
        </button>
        {{else}}
        <button type="button" data-oid="{{oid}}" aria-label="Deploy this commit to {{../target}}."
          class="deploy btn btn-outline tooltipped tooltipped-n"
          name="button">
          {{> icon_play}}
        </button>
        {{/if}}
        {{/if}}
      </div>
      {{/each}}
    </div>
  </div>
</div>

<script>
  (function () {
    var deploys = document.getElementsByClassName("deploy")
    for (var i = 0; i < deploys.length; i++) {
      deploys[i].onclick = function (e) {
        var sha = this.dataset.oid;
        e.preventDefault();
        fetch("/deploy/{{target}}/{{owner}}/{{repo}}/" + sha, { method: "POST" })
          .then(function (resp) { return resp.json() })
          .then(function (data) {
            if (data.error) {
              error(data.error);
              return;
            }
            window.location.href = "/logs/{{owner}}/{{repo}}/" + data.id;
          }).catch(function (err) {
            error("Failed to create deployment.");
          });
      }
    }
  })()

  function error(message) {
    var container = document.getElementById("error-container");
    var div = document.createElement("div");
    div.className = "flash flash-error";
    div.innerText = message;
    container.appendChild(div);

    setTimeout(function() {
      div.remove();
    }, 5000);
  }

  (function () {
    var active;
    function toggle(id) {
      if (active) {
        dismiss(active);
      }
      if (active === id) {
        active = null;
        return;
      }
      active = id;
      var el = document.getElementById(id);
      el.classList.add("d-block");
      el.classList.remove("d-none");
    }
    function dismiss(id) {
      var el = document.getElementById(id);
      el.classList.add("d-none");
      el.classList.remove("d-block");
    }
    var toggles = document.querySelectorAll('[data-toggle]');
    for (var i = 0; i < toggles.length; i++) {
      toggles[i].addEventListener('click', function(e) {
        e.preventDefault();
        toggle(this.dataset.toggle);
      });
    }
  })();
</script>

{{> footer}}
