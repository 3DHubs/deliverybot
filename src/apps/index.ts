import { auth } from "./auth";
import { deploy } from "./deploy";
import { home } from "./home";
import { commands } from "./commands";

export const apps = [home, auth, deploy, commands];
