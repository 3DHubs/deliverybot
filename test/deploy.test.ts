import { View } from "../src/apps/deploy"


describe("DeployView", () => {
  it("renders a payload", () => {
    const query = require("./fixtures/query.json")
    expect(View("deliverybot", "example", "master", query.data)).toMatchSnapshot();
  })
})
