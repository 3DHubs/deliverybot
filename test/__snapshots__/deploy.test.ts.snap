// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeployView renders a payload 1`] = `
Object {
  "branches": Array [
    Object {
      "active": false,
      "name": "colinjfw-patch-1",
    },
    Object {
      "active": false,
      "name": "colinjfw-patch-2",
    },
    Object {
      "active": false,
      "name": "colinjfw-patch-3",
    },
    Object {
      "active": false,
      "name": "colinjfw-patch-4",
    },
    Object {
      "active": true,
      "name": "master",
    },
  ],
  "commits": Array [
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployments": Array [],
      "message": "Update deploy.yml",
      "oid": "244a445",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "https://console.cloud.google.com/cloud-build/builds/01ecc3eb-1a88-428a-ba0a-d421c90eb360?project=deliverybot",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "f28d263",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "red",
          "failure": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "red",
          "failure": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "https://console.cloud.google.com/cloud-build/builds/b68f76c5-31f0-47e7-a28d-a1bebec9fc39?project=deliverybot",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "03938e0",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "red",
          "failure": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "gray",
            "waiting": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "gray",
            "waiting": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "yellow",
            "pending": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": null,
        },
      ],
      "message": "Update deploy.yml",
      "oid": "266a098",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "gray",
          "waiting": true,
        },
      },
      "deployments": Array [],
      "message": "Update deploy.yml",
      "oid": "13b08bb",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161479032",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "fc245f7",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161440765",
        },
        Object {
          "creator": "colinjfw",
          "environment": "staging",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161440865",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "67504c8",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161440458",
        },
        Object {
          "creator": "colinjfw",
          "environment": "staging",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161440575",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "fd3d28a",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "green",
          "success": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "red",
          "failure": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/89cf477a-f7d0-4a84-948e-ee8fa535f5b4",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/32ddb9ab-27cb-415c-94ce-461d843cd7d3",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/de3f0b69-cbd5-4f52-a8de-5d73167619d1",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/317ab83f-80ee-474e-8e09-38887ee79194",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/52b74663-c848-497b-b52b-ca5425b035e3",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "red",
            "failure": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/63a8cc8c-ed24-4d79-b97f-0c70ce6a8f9e",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/2e9ec0e5-98d9-4b1c-86e7-2086d3764be8",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161324570",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161324837",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161325022",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161325741",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161337388",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161337565",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161338055",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161338198",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161338704",
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "green",
            "success": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/161345711",
        },
      ],
      "message": "Update deploy.yml",
      "oid": "eb7930e",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
    Object {
      "author": "colinjfw",
      "check": Object {
        "status": Object {
          "color": "yellow",
          "pending": true,
        },
      },
      "deployment": Object {
        "status": Object {
          "color": "yellow",
          "pending": true,
        },
      },
      "deployments": Array [
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "gray",
            "waiting": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "production",
          "status": Object {
            "color": "gray",
            "waiting": true,
          },
          "url": null,
        },
        Object {
          "creator": "colinjfw",
          "environment": "k8s",
          "status": Object {
            "color": "yellow",
            "pending": true,
          },
          "url": "http://localhost:3000/logs/deliverybot/example/c94697a5-dc47-4c7f-b632-2240ae8b7d34",
        },
      ],
      "message": "Create deploy.yml",
      "oid": "626285e",
      "owner": "deliverybot",
      "repo": "example",
      "undeployed": false,
    },
  ],
}
`;
